### PR TITLE
Distributed planning evaluation framework with GCP

### PR DESCRIPTION
**Pull request by riemanli**
# Instruction of setup
## Requirements
* To be able to access the GCS bucket, make sure you have the read/write permissions of the target bucket in your GCP project.
    Check out the [how-to guide in Google Cloud Storage](https://cloud.google.com/storage/docs/how-to) to learn more about GCS bucket.
* Install Cloud SDK following [the instruction](https://cloud.google.com/sdk/docs/install). Then, to authenticate the terminal session with your google account, run the following commands in your terminal:
    * ```gcloud auth login```
    * ```gcloud auth application-default login```

    Log in with your Google account that is granted with read/write permissions in the pop-up browser. Note that the permission is granted only for the current terminal session. If the session is closed, the same commands need to be run again.
* Please follow the [steps](https://cloud.google.com/dataflow/docs/quickstarts/quickstart-python#get-the-apache-beam-sdk) to install the Apache Beam SDK with Dataflow support in your virtual environment.
* To use the non-PyPI dependency in Dataflow mode, [`cardinality_estimation_evaluation_framework`](https://github.com/world-federation-of-advertisers/cardinality_estimation_evaluation_framework), we need to create a tarball file of it, so we can pass it to Dataflow for installation. To get the tarball, in your terminal session, `cd` to the directory that contains `setup.py` of the `cardinality_estimation_evaluation_framework` repo and run the command:

    ```
    ​python setup.py sdist
    ```

    Please see more details in this [document](https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/#local-or-nonpypi) about why a tarball is needed.

## Execution Command
* Running the whole evaluation is done by running the `main` function in [experiment_driver.py](https://github.com/world-federation-of-advertisers/planning-evaluation-framework/blob/rieman_distributed_evaluation/src/driver/experiment_driver.py) by executing this [command](https://github.com/world-federation-of-advertisers/planning-evaluation-framework/blob/rieman_distributed_evaluation/src/driver/experiment_driver.py#L19) in the terminal.

### CODE REVIEW

File: src/driver/tests/experimental_design_test.py [hunk 1]
```diff
@@ -190,6 +228,118 @@ def test_evaluate_with_two_cores(self):
...
+
+        with TestPipeline() as p:
+            output = (
+                p
+                | beam.Create([FakeExperimentalTrial(i) for i in range(num_trials)])
+                | beam.ParDo(EvaluateTrialDoFn(), seed)
+            )
+            assert_that(output, equal_to(expected))
+
+    def test_combine_dataFrame_fn(self):
```
Comment (matthewclegg) [inline src/driver/tests/experimental_design_test.py:327]: Please turn off all of the log messages that this function is generating to stdout.

File: src/driver/tests/experimental_design_test.py [hunk 2]
```diff
@@ -190,6 +228,118 @@ def test_evaluate_with_two_cores(self):
...
+                "--runner=direct",
+                "--direct_running_mode=multi_processing",
+                f"--temp_location={experiment_dir}",
+                f"--direct_num_workers={num_workers}",
+            ]
+        )
+        pipeline_options = PipelineOptions(pipeline_args)
+        return exp.load(use_apache_beam=True, pipeline_options=pipeline_options)
+
+    def test_evaluate_trial_do_fn(self):
```
Comment (matthewclegg) [inline src/driver/tests/experimental_design_test.py:314]: Please turn off all of the log messages that this function is generating to stdout.

File: src/driver/tests/experimental_design_test.py [hunk 3]
```diff
@@ -190,6 +228,118 @@ def test_evaluate_with_two_cores(self):
...
+            self.assertEqual(results.shape[0], 1)
+            self.assertAlmostEqual(results["relative_error_at_100"][0], 0.0, delta=0.01)
+            self.assertGreater(results["max_nonzero_frequency_from_halo"][0], 0)
+            self.assertEqual(results["max_nonzero_frequency_from_data"][0], 1)
+
+    @patch.object(experimental_design, "GSClient", LocalGSClient)
+    @patch.object(experimental_design, "EvaluateTrialDoFn", FakeEvaluateTrialDoFn)
+    @patch.object(data_set, "GSPath", LocalGSPath)
+    @patch.object(data_design, "GSPath", LocalGSPath)
+    def test_evaluate_with_apache_beam_with_cloud_path(self):
```
Comment (matthewclegg) [inline src/driver/tests/experimental_design_test.py:249]: Please turn off all of the log messages that this function is generating to stdout.

File: src/driver/tests/experiment_driver_test.py [hunk 4]
```diff
@@ -18,7 +18,15 @@
 import pandas as pd
```
Comment (matthewclegg) [inline src/driver/tests/experiment_driver_test.py:18]: When I run this test on my local machine, it now hangs after completing all tests.

**Comment by riemanli** [comment]
Hey @matthewclegg, thanks for the feedback! Please see the following as the response:
* > Add a step at the beginning instructing the user to download and install the Cloud SDK, and give a link to the page for doing so.

    The installment should be done if the steps mentioned in the instruction are done:

     > Please follow the [steps](https://cloud.google.com/dataflow/docs/quickstarts/quickstart-python#get-the-apache-beam-sdk) to install the Apache Beam SDK with Dataflow support in your virtual environment.

    Basically, ```pip install 'apache-beam[gcp]'``` mentioned in the hyperlink will install Apache Beam and required GCP dependencies. Please let me know if the installment doesn't work for you to get the Cloud SDK, and if you want me to explicitly list ```pip install 'apache-beam[gcp]'``` in the instruction.

* > Add link to page that explains about GCS buckets and how to create/use them.

    Thanks for the suggestion. The link is added.

* > The document that was linked regarding creating a tarball for the cardinality estimation framework doesn't seem to say anything about creating tarballs.
    
    In the linked document, it said:
    >  If you have the `setup.py` for that package then you can build the tarball with the following command:
        ``` ​python setup.py sdist```
    
    which I think should provide the information of creating tarballs. Please let me know if that is not clear to you.

File: src/driver/tests/experiment_driver_test.py (hunk 4 above)
Comment (riemanli) [reply src/driver/tests/experiment_driver_test.py:18]: Thanks for catching that. It turns out that `pathos 0.2.5` causes this issue. `pathos 0.2.7` doesn't have it, though it has dependency conflict with Apache Beam. I will modify `requirements.txt` and `setup.py` to use `pathos 0.2.7` and change the instruction of setup.

Comment (riemanli) [reply src/driver/tests/experiment_driver_test.py:18]: Added the instruction of the setup and the execution example to `README`.

File: README.md [hunk 5]
```diff
@@ -77,10 +76,35 @@ To install the dependencies, run:
...
+pip install apache-beam[gcp]
 ```
 
-To execute the Planning Evaluation Framework, we have found it most convenient to
-modify PYTHONPATH and create a symbolic link to the source code, following these steps:
+Note that there is a dependency conflict between Apache Beam and Pathos. The current workaround is manually installing Apache Beam after installing the packages in `requirements.txt`, so the dependencies of Apache Beam can be satisfied.
+
+To utilize the support of distributed computing using GCP Dataflow, the following requirements have to be fulfilled beforehand:
+1. Make sure to have read/write permissions of the target Google Cloud Storage (GCS) bucket in your GCP project. Check out the [how-to guide in Google Cloud Storage](https://cloud.google.com/storage/docs/how-to) to learn more about GCS bucket.
+2. Install Cloud SDK by following the [instruction](https://cloud.google.com/sdk/docs/install). Before running the program with Dataflow, authenticate the terminal session with your google account by running the following commands in your terminal:
```
Comment (jiayu-google) [inline README.md:86]: Sorry, I'm stuck at this step.  After installing cloud SDK, I ran `gloud auth login` but always got `bash: gcloud: command not found`.  Under which dir shall I run the command?  tried several ones but they didn't work.

File: README.md [hunk 6]
```diff
@@ -77,10 +76,35 @@ To install the dependencies, run:
 
 ```
 pip install -r requirements.txt
+pip install apache-beam[gcp]
```
[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:79]: Yes. This is expected as there is a dependency conflict between `pathos` and `apache-beam`. 

    apache-beam 2.31.0 depends on dill<0.3.2 and >=0.3.1.1
    pathos 0.2.7 depends on dill>=0.3.3

To resolve the conflict, `pathos-0.2.5`, which requires `dill>=0.3.1`, was needed. However, `pathos-0.2.5` has critical bugs that are not acceptable, so the current workaround is manually installing Apache Beam after installing the packages in `requirements.txt`, so the dependencies of Apache Beam can be satisfied, and Pathos can run without issues.

If you run the `pip install` command in the instructed sequence, you should be able to obtain all necessary dependencies to run the program, though there is an error as you mentioned.

File: README.md (hunk 5 above)
Comment (riemanli) [reply README.md:86]: Which instruction did you use? If you are using Cloudtop, you need to follow the instruction in the `Linux` tab.

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -156,15 +179,42 @@ python3 driver/experiment_driver.py \
...
-recorded in a CSV file named `$DIR/results`.  You can then load this
-into colab and explore the results that were obtained.  For some
-example colabs, see the `analysis` directory.
+Setting `cores=0` enables multithreading on all available cores.  
+
+In addition, you may use Apache Beam to evaluate the Experiments in two different modes -- direct runner and Dataflow runner. For the direct runner mode, use multi-processing with the execution command:
+
+```
+python3 driver/experiment_driver.py \
+  --data_design_dir=$DIR/data \
```
[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:188]: Hi Matthew, thanks for catching that. I just realized I missed `--` in the example command. The correct command should be:

```diff
+ python3 driver/experiment_driver.py -- \
- python3 driver/experiment_driver.py \
[arguments]...
```

[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:188]: Fixed the example execution commands in the instruction.

**Review by jiayu-google — COMMENTED** [review]
Almost there!!

File: README.md [hunk 8]
```diff
@@ -146,25 +170,48 @@ M3 model.
...
+  --experimental_design=driver/single_publisher_design.py \
+  --output_file=gs://<bucket_name>/<subpath/to/output_file> \
+  --intermediates_dir=gs://<bucket_name>/<subpath/to/intermediates_dir> \
+  --use_apache_beam \
+  --runner=DataflowRunner \
+  --region=<region> \
+  --project=<gcp_project_id> \
+  --staging_location=gs://<bucket_name>/<subpath/to/staging_dir> \
+  --setup_file=<path/to/setup.py> \
+  --extra_package=<path/to/wfa_cardinality_estimation_evaluation_framework-0.0.tar.gz>
```
[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:211]: Thanks for the suggestion. Modified the README so it mentioned `wfa_cardinality_estimation_evaluation_framework-0.0.tar.gz` explicitly.

File: README.md [hunk 9]
```diff
@@ -146,25 +170,48 @@ M3 model.
...
+```
+
```
[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:213]: It seems like that Python 3.9.7 is not officially supported by Dataflow. Could you try again with adding `--experiment use_unsupported_python_version` as instructed in the last line of the error message?

[the remark this answers is not in the document]
Comment (riemanli) [reply README.md:213]: Also, an alternative is using Python 3.8.

[the remark this answers is not in the document]
Comment (jiayu-google) [reply README.md:213]: Thanks Rieman!
1. some error with `experiment use_unsupported_python_version`. [see pages 1-2 of this doc](https://docs.google.com/document/d/1IOXGUIYUZ3ZuHBgexlZtPNJWEBM5a1xCrASdm7EOV74/edit?resourcekey=0-3JBqpClvNgadu63OTxtaTA) for details.
2. Sorry for the dumb question, could you teach me how to switch to python 3.8 given that I have 3.9 in the current virtual env?

**Review by jiayu-google — COMMENTED** [review]
Thank you for your quick reply and actions!

**Comment by riemanli** [comment]
Investigated the usage of Python3.9. Based on the truth that Python3.9 is not officially supported by the Dataflow runner and on the test that the program would be stuck with Python3.9 in the Dataflow mode, we highly recommend using Python3.8 for this repo.

**Review by jiayu-google — APPROVED** [review]
This is absolutely important and difficult work.  Thank you for your tremendous efforts, Rieman.

**Review by amiragha — COMMENTED** [review]
There isn't a lot methodology involved here and I am not an expert on this commit. As for me, if it runs, I am fine with it :)

**Pull request merged**